op-batcher/batcher/driver.go

```diff
@@ -635,32 +645,31 @@ func (l *BatchSubmitter) waitNodeSync() error {
 }
 
 // publishStateToL1 queues up all pending TxData to be published to the L1, returning when there is no more data to
-// queue for publishing or if there was an error queing the data.  maxDuration tells this function to return from state
-// publishing after this amount of time has been exceeded even if there is more data remaining.
-func (l *BatchSubmitter) publishStateToL1(queue *txmgr.Queue[txRef], receiptsCh chan txmgr.TxReceipt[txRef], daGroup *errgroup.Group, maxDuration time.Duration) {
-	start := time.Now()
+// queue for publishing or if there was an error queing the data.
+func (l *BatchSubmitter) publishStateToL1(ctx context.Context, queue *txmgr.Queue[txRef], receiptsCh chan txmgr.TxReceipt[txRef], daGroup *errgroup.Group) {
 	for {
-		// if the txmgr is closed, we stop the transaction sending
-		if l.Txmgr.IsClosed() {
-			l.Log.Info("Txmgr is closed, aborting state publishing")
-			return
-		}
-		if !l.checkTxpool(queue, receiptsCh) {
-			l.Log.Info("txpool state is not good, aborting state publishing")
+		select {
+		case <-ctx.Done():
 			return
-		}
+		default:
+			// if the txmgr is closed, we stop the transaction sending
+			if l.Txmgr.IsClosed() {
+				l.Log.Info("Txmgr is closed, aborting state publishing")
+				return
+			}
+			if !l.checkTxpool(queue, receiptsCh) {
+				l.Log.Info("txpool state is not good, aborting state publishing")
+				return
+			}
 
-		err := l.publishTxToL1(l.killCtx, queue, receiptsCh, daGroup)
+			err := l.publishTxToL1(l.killCtx, queue, receiptsCh, daGroup)
 
-		if err != nil {
-			if err != io.EOF {
-				l.Log.Error("Error publishing tx to l1", "err", err)
+			if err != nil {
+				if err != io.EOF {
+					l.Log.Error("Error publishing tx to l1", "err", err)
+				}
+				return
 			}
-			return
-		}
-		if time.Since(start) > maxDuration {
-			l.Log.Warn("Aborting state publishing, max duration exceeded")
-			return
```

Comment:
We don't need to return from this loop when we "run out of time" any more, since throttling and reading blocks now happen concurrently.